### DISCUSSION TITLE
Thoughts about calibration and fixed starting position

### DISCUSSION THREAD

**Discussion by thomasbrus** [pybricks/support#1828] [Q&A]
Let's say I am building a robot arm. I may use a worm gear to reduce the gear ratio and to make sure the weight of the arm doesn't put load on the motor in stationary position.

At the start of my program, I would like the arm to start in a fixed/known position. For two reasons:

- To know how far I can rotate back and forth before hitting a limit
- If I know the position of the arm at the start, I can apply inverse kinematics to move the robot arm to an absolute x, y, z coordinate.

The usual approach is to drive the motor back and forth using  [run_until_stalled](https://docs.pybricks.com/en/stable/pupdevices/motor.html), that works well for calibrating the steering of a car for example. But considering the weight of the arm and that will be using a worm gear, that is not really an option.

The angular motors also know their zero position, but that is only useful when the gear ratio is <1. If the motor needs to rotate many times (when using a worm gear for example), then resetting it to zero position doesn't mean the arm is in the zero position as well (not necessarily).

Are there any other calibration techniques that I haven't thought about? I suppose I could use touch sensors, but that solution has downsides as well. For now I just settle on manually putting the arm in the right position and returning to the initial position at the end of the program, or when it crashes.

**Comment by Debenben** [comment]
An idea could be to use acceleration sensors to get the angle with respect to gravity. I am using the differences in acceleration of technichubs and (WeDo-) tilt sensors (using device.read(3)) to calculate leg positions of my lego robot dog. The dog is not finished but getting joint angles this way works.

**Reply by thomasbrus** [reply]
Nice, I'm also working on a quadruped, same calibration problem there. Can you explain it in a bit more detail? How can you determine the initial angles exactly using the acceleration?

**Reply by Debenben** [reply]
It looks like this (sorry for bad image quality, currently that is all I have):
![IMG_20240806_124245](https://github.com/user-attachments/assets/861bea6e-01d9-409a-8ee5-0170fd9d510f)

Each leg has 3 joints with 3 motors. Two motors are connected to the technic hubs in the body. The third is connected to a technic hub in the upper part of the leg. In the lower part of the leg there is a wedo-tilt sensor. For easier maintenance it is possible to remove the large black gear on the outside and move it by hand.

I have put the code on github (sorry also for bad code quality, I will clean it up once it is finished)

In general it is using the "MotorAngle" values of the motor encoder and can convert it to the angles of the joints and back:
https://github.com/Debenben/babyele/blob/1dce007d2da70a9035fc01004ee20370a786476b/src/conversions.ts#L26
https://github.com/Debenben/babyele/blob/1dce007d2da70a9035fc01004ee20370a786476b/src/conversions.ts#L36

The acceleration is used for calibration after start, validating and recalibrating due to mechanical play of gears. The conversion function is this:
https://github.com/Debenben/babyele/blob/1dce007d2da70a9035fc01004ee20370a786476b/src/conversions.ts#L140
It is working well in "normal" positions. In case a rotation axis is parallel to gravity it does not work. There the angle would be the difference in absolute heading direction which you cannot measure with accelerometers. The values get more and more inaccurate the closer you get to that position. I simply refrain from recalibrating in such a situation and use motor encoders only.

The knee took a lot of redesigns to be able to handle the weight and deal with the cables running through it but it works well now. I did not want to use linear actuators there, but they are definitely the easier option when dealing with larger weights. The top joint with the possibility to infinitely rotate sometimes skips gears under load, that I still would like to redesign.

Another problem is communication with the hubs. I tried conneting them with the lego firmware to one computer with a bluetooth dongle, but I only have bluetooth dongles which refuse to connect to 6 hubs at the same time or they allow to connect 6 but then randomly disconnect a hub when trying to send moderate amounts of messages. With pybricks I can use the observe and broadcast which does not have this problem, but unfortunately for technic hubs there is still a bug where they sometimes randomly get stuck when broadcasting and observing at the same time.

**Reply by thomasbrus** [reply]
Looks really cool. I understand the part about converting between motor angles and leg positions (kinematics). Not really sure what is going on [here though](https://github.com/Debenben/babyele/blob/1dce007d2da70a9035fc01004ee20370a786476b/src/conversions.ts#L140-L156) 🤔 how does this accomplish getting the legs in the right position from an unknown starting position?

**Reply by Debenben** [reply]
The acceleration gets you pan and tilt angle with respect to gravity. Lets assume I measure some acceleration values x, y, and z in the hub in the body of the dog and then the same values in the hub in the upper part of a leg. Then the hub which is in the leg must have the same orientation to gravity. In my case the hubs in the legs are mounted with the light pointing down and facing to the outside, therfore I know if x, y, and z of the hub in body and upper part of the leg are the same, my top part of the leg must be pointing forward at 90 degrees and tilted to the side by 90 degrees.

If I know my gear ratio to be 100:1, the position of both my motors at the top and must be 9000 degrees. Then I would set them to 9000 and request them to move to 0 if I want them to move to my "starting position".

The same you can do comparing acceleration in upper and lower part of the leg. In this case I have one degree of freedom the joint so it is overdetermined and I would have the possibility to check that the legs do not bend where they should not. In my case the devices are mounted such that the tilt angles they measure must always be the same. Therefore I can directly subtract the two pan angles for top and bottom hub which I get with arctan2(z, x) from the acceleration in z and x direction.

**Reply by thomasbrus** [reply]
Okay I get how you can compare the pan and tilt between the hub in the body and the leg to determine the leg position and thus the motor angles, that makes sense yes 👍  you can also connect the hubs via the Bluetooth Web API by the way, in case you're interested in that I [have some code over here](https://github.com/thomasbrus/lego-swerve-drive/tree/four-wheel-swerve-drive/src/hooks).

**Reply by thomasbrus** [reply]
@Debenben do you happen to have more pictures (or an explanation) of how the knee joint works? I can't really tell from the picture if there is a turntable inside of it or if it is connected via an axle. Really cool build.

**Reply by Debenben** [reply]
@thomasbrus sure. I don't have the model with me, but I can make a drawing:

![knee](https://github.com/user-attachments/assets/457fc814-38c8-453e-a265-a7bd6dc7368e)

[knee.ldr.txt](https://github.com/user-attachments/files/18400722/knee.ldr.txt)

The gear rack is attached to the upper part of the leg. The liftarm construction in the middle holds the two sides of the upper part together and is fixed with axles that can be removed for easy maintenance. 3 extension cables (bought from aliexpress) for motor, tilt sensor and color-distance-sensor run through the middle, they are hidden behind the yellow panel in the first photo. Everything you see in the drawing is mirrored in the middle of the yellow axles.
When I find some time I can complete the model, I hope this small sketch already helps and maybe you find an even better construction technique than me. 

I have also rebuild the top part of the leg using one gear-rack and one turntable now. If you are interested I can try to also do a model of that.

**Reply by thomasbrus** [reply]
Ahhh those gear rack pieces, makes sense, I see it now. Only have two of those... That must be pretty strong. I hadn't thought of that. Yes would also like to see the top part :) Am using single 60 tooth turn tables for that now. Was trying out some things myself as well for the lower part as you can see in the prototype below. Idea is that it's fine if the knee rotation is limited, and to use a linear actuator (I also don't have these pieces..) or worm gears instead 😄 

[worm_gear_leg.io.txt](https://github.com/user-attachments/files/18401858/worm_gear_leg.io.txt)
![worm_gear_leg_2](https://github.com/user-attachments/assets/d34bfa02-fbf2-4a6b-9076-894a88d32177)

**Reply by thomasbrus** [reply]
@Debenben How's it going with your quadruped project? 😄

**Reply by Debenben** [reply]
@thomasbrus 
I redesigned the upper part of the leg a little. Movement was painfully slow and whenever I tried to make it faster it did not have enough momentum to carry the weight. To improve it, my idea is to shorten the legs by making the hub move to one side and making it asymetric such that it can bend to one side much further than to the other. The upper part of the front legs designed with the new concept, the ones in the back is still the old:
![quadruped](https://github.com/user-attachments/assets/ab0599c3-52de-4af0-a207-6495283cfac0)
Next up would be to give it some final design touches, copy the new design to the rear legs and do a similar change to the lower part, but it will take some time I guess.

How is your quadruped project going?

**Reply by thomasbrus** [reply]
Love the design and the looks of it!

> I redesigned the upper part of the leg a little. Movement was painfully slow and whenever I tried to make it faster it did not have enough momentum to carry the weight

Yes it is quite difficult to get this balance right.  It also doesn't sit right with me to put too much stress on Lego parts 😅

> How is your quadruped project going?

Well, it's on pause now but I tried many different designs. The issue was that they were to heavy to lift itself. 

Made a dashboard to visualize the (inverse) kinematics. It works well, but haven't implemented any type of bluetooth communication yet though.

Last video/photo is the latest design. Yes, it's quite far from what a quadruped typically looks like🤣 Used up all my parts for this haha. But most importantly, it can comfortably lift itself. If I have some time again I'm gonna look at implementing a walking sequence and all that.


Failed design:

https://github.com/user-attachments/assets/9af56d94-b2b8-4f48-b0be-23ca20a75d4f


![WhatsApp Image 2025-10-05 at 14 49 22](https://github.com/user-attachments/assets/257bb2d5-2b21-4423-a049-d9dc2f17d2b9)

---

Latest attempt:

![WhatsApp Image 2025-10-05 at 00 15 32](https://github.com/user-attachments/assets/ebcf4be5-9cfa-4032-91d5-362c10ee6757)

https://github.com/user-attachments/assets/4556faf4-0d7f-4277-8f96-577952abb3e0

**Comment by BertLindeman** [comment]
Thomas, 

I have it not at hand now, But I had a tic-tac-toe robot, where I had a smooth calibration function by lowering the torque.
At the moment no time to look for it. So later......

Bert

**Comment by laurensvalk** [comment]
This is a partial answer, but it is simple and practical. It does not require any extra sensors. I had the same challenge for this [6-legged robot](https://www.youtube.com/watch?v=ArybyRRnqQs).

When you initialize the motor, you have the option of `reset_angle=False`, so it will maintain the angle.

So then you can manually initialize once - for example the first time after you boot - and use `reset_angle(0)` or with some sensible value. And then you don't need calibration again while the hub is on. And you can restart your program many times or change your program, and still maintain the position.

In this robot, I made it choose `reset_angle=True` if I held one of the buttons when the program was started, and `False` otherwise.

**Reply by thomasbrus** [reply]
Sweet! So for how long does it "remember"  the angle? How does that work exactly? Apparently it works even when you upload a new version of the program? Thanks, gonna give that a try. To calibrate it once manually is perfectly fine.

**Reply by laurensvalk** [reply]
The motors always keep counting, even when the program ends. By choosing `reset_angle=False`, it just doesn't "modulo" it back to the `-180` to `180` range.

So the only logic you'd need to add is something like:

```python
should_reset = some_condition_like_left_button_pressed

if should_reset:
   beep
   # here you can manually put the legs in the right place
   wait_for_some_other_condition_so_you_can_indicate_completion
   double beep

motor = Motor(Port.A, reset_angle=should_reset)

```

**Reply by thomasbrus** [reply]
Gotcha. Should be fairly easy to do with a button press on an xbox controller for example 😄

**Reply by thomasbrus** [reply]
Ended up making a separate script for motor calibration

```python
# calibrate.py

from pybricks.hubs import TechnicHub
from pybricks.parameters import Port, Button, Color
from motors import MaybeMotor, NoneMotor
from pybricks.iodevices import XboxController

hub = TechnicHub()
hub.light.on(Color.RED)

controller = XboxController()

hub.light.off()
controller.rumble(50)


def set_corresponding_light(index):
    if index == 0:
        hub.light.on(Color.GREEN)
    elif index == 1:
        hub.light.on(Color.RED)
    elif index == 2:
        hub.light.on(Color.BLUE)
    elif index == 3:
        hub.light.on(Color.YELLOW)


motor_a = MaybeMotor(Port.A)
motor_b = MaybeMotor(Port.B)
motor_c = MaybeMotor(Port.C)
motor_d = MaybeMotor(Port.D)
motors = [motor_a, motor_b, motor_c, motor_d]

index = 0

while index < len(motors) and isinstance(motors[index], NoneMotor):
    index += 1

set_corresponding_light(index)

while True:
    pressed_buttons = controller.buttons.pressed()

    if Button.GUIDE in pressed_buttons:
        for motor in motors:
            motor.reset_angle(0)
        hub.system.shutdown()

    left, right = controller.triggers()
    motors[index].dc(right - left)

    if Button.A in pressed_buttons:
        index = 0
        set_corresponding_light(index)

    if Button.B in pressed_buttons:
        index = 1
        set_corresponding_light(index)

    if Button.X in pressed_buttons:
        index = 2
        set_corresponding_light(index)

    if Button.Y in pressed_buttons:
        index = 3
        set_corresponding_light(index)

```

```python
# See https://pybricks.com/project/disable-motor-detect/
# motors.py

from pybricks.pupdevices import Motor
from pybricks.parameters import Direction


# A function or method that just does nothing.
def do_nothing(*args, **kwargs):
    return 0


# A class that does nothing and allows any method calls.
class NoneMotor:
    def __getattr__(self, name):
        return do_nothing


# Helper function that creates a Motor-like object, but
# suppresses exceptions about motors not being connected.
def MaybeMotor(port, direction=Direction.CLOCKWISE):
    try:
        # Give a normal motor object if such device is connected.
        return Motor(port, direction)
    except OSError:
        return NoneMotor()
```

**Comment by BertLindeman** [comment]
I see more interesting answers here. 

So I just post a link to my [tic-tac-toe program on bricksafe](https://bricksafe.com/pages/Bert_bricksafe/moc-129823-tictactoe/pybricks-programs)

Bert

**Reply by thomasbrus** [reply]
Thanks for sharing the code. I have gone down a similar path in another project (although it was much simpler than your code), but always felt there had to be a more elegant way that doesn't stress the gears and motors as much. That being said it is nice that the calibration is entirely autonomous!
